### PR TITLE
Reland '[hwasan] Add fixed_shadow_base flag (#73980)'

### PR DESCRIPTION
**Pull request by thurstond**
This was reverted in https://github.com/llvm/llvm-project/pull/95435 because it broke Android static hwasan binaries. This reland limits the change to !SANITIZER_ANDROID.

Original commit message:
When set to non-zero, the HWASan runtime will map the shadow base at the specified constant address.

This is particularly useful in conjunction with the existing compiler option 'hwasan-mapping-offset', which bakes a hardcoded constant address into the instrumentation.

### CODE REVIEW

File: compiler-rt/lib/hwasan/hwasan_linux.cpp [hunk 1]
```diff
@@ -106,8 +106,13 @@ static uptr GetHighMemEnd() {
 }
 
 static void InitializeShadowBaseAddress(uptr shadow_size_bytes) {
-  __hwasan_shadow_memory_dynamic_address =
-      FindDynamicShadowStart(shadow_size_bytes);
+  // Android static HWASan doesn't initialize flags correctly
+  if (!SANITIZER_ANDROID && flags()->fixed_shadow_base != (uptr)-1) {
```
Comment (vitalybuka) [inline compiler-rt/lib/hwasan/hwasan_linux.cpp:110]: please use SANITIZER_CAN_USE_PREINIT_ARRAY
as it's the real rootcause

**Comment by fmayer** [comment]
If you can wait, I can confirm this works later today by cherry-picking this and trying to reproduce the Android problem.

**Comment by vitalybuka** [comment]
> If you can wait, I can confirm this works later today by cherry-picking this and trying to reproduce the Android problem.

LGTM but not critical, as the cause looks straightforward.

However fix for flag initialization will need real testing.

File: compiler-rt/lib/hwasan/hwasan_linux.cpp (hunk 1 above)
Comment (thurstond) [reply compiler-rt/lib/hwasan/hwasan_linux.cpp]: Done

**Review by vitalybuka — APPROVED** [review]

**Comment by fmayer** [comment]
Now also on the correct PR:

I think that's incorrect:

```
#ifndef SANITIZER_CAN_USE_PREINIT_ARRAY
#if (SANITIZER_LINUX || SANITIZER_FUCHSIA || SANITIZER_NETBSD) && !defined(PIC)
#define SANITIZER_CAN_USE_PREINIT_ARRAY 1
// Before Solaris 11.4, .preinit_array is fully supported only with GNU ld.
// FIXME: Check for those conditions.
```

ANDROID is Linux and does in fact _support_ preinit array, we just don't use it for HWASan.

**Comment by vitalybuka** [comment]
> Now also on the correct PR:
> 
> I think that's incorrect:
> 
> ```
> #ifndef SANITIZER_CAN_USE_PREINIT_ARRAY
> #if (SANITIZER_LINUX || SANITIZER_FUCHSIA || SANITIZER_NETBSD) && !defined(PIC)
> #define SANITIZER_CAN_USE_PREINIT_ARRAY 1
> // Before Solaris 11.4, .preinit_array is fully supported only with GNU ld.
> // FIXME: Check for those conditions.
> ```
> 
> ANDROID is Linux and does in fact _support_ preinit array, we just don't use it for HWASan.

I believe it's disabled because hwasan is dynamic lib.
Either way this patch is correct for the current state of the code.

**Comment by fmayer** [comment]
> > Now also on the correct PR:
> > I think that's incorrect:
> > ```
> > #ifndef SANITIZER_CAN_USE_PREINIT_ARRAY
> > #if (SANITIZER_LINUX || SANITIZER_FUCHSIA || SANITIZER_NETBSD) && !defined(PIC)
> > #define SANITIZER_CAN_USE_PREINIT_ARRAY 1
> > // Before Solaris 11.4, .preinit_array is fully supported only with GNU ld.
> > // FIXME: Check for those conditions.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ANDROID is Linux and does in fact _support_ preinit array, we just don't use it for HWASan.
> 
> I believe it's disabled because hwasan is dynamic lib. Either way this patch is correct for the current state of the code.

I'll double check but I don't think it is, the bug still happened on Android when I applied this

**Comment by vitalybuka** [comment]
Oh, `#if (SANITIZER_LINUX ||` applies to android, so we need original version of the patch.

**Comment by vitalybuka** [comment]
@fmayer Can you please check the 8b6c034 ?

**Comment by fmayer** [comment]
> SANITIZER_CAN_USE_PREINIT_ARRAY



> @fmayer Can you please check the [8b6c034](https://github.com/llvm/llvm-project/commit/8b6c0349b8885e5183d51d98c6d30927b978e842) ?

Works

**Review by fmayer — APPROVED** [review]

File: compiler-rt/lib/hwasan/hwasan_linux.cpp [hunk 2]
```diff
@@ -106,8 +106,12 @@ static uptr GetHighMemEnd() {
 }
 
 static void InitializeShadowBaseAddress(uptr shadow_size_bytes) {
-  __hwasan_shadow_memory_dynamic_address =
-      FindDynamicShadowStart(shadow_size_bytes);
+  if (!SANITIZER_ANDROID && flags()->fixed_shadow_base != (uptr)-1) {
```
Comment (fmayer) [inline compiler-rt/lib/hwasan/hwasan_linux.cpp:109]: Please add a comment why we are doing this here.

**Comment by vitalybuka** [comment]
> > @fmayer Can you please check the [8b6c034](https://github.com/llvm/llvm-project/commit/8b6c0349b8885e5183d51d98c6d30927b978e842) ?
> 
> Works

Thanks, testing was very helpful!

**Comment by thurstond** [comment]
Thanks @fmayer for testing and @vitalybuka for landing the appropriate patch!

**Pull request merged**